Support --all-namespaces option for service and revision list

nak3 opened the pull request:
Currently `kn {service, revision} list` does not have the all-namespaces
options. Although it is possible to list them by `kn service list -n
""`(empty string), it is difficult to find such secret trick.

Hence, this patch adds `--all-namespaces` option.

navidshaikh commented on pkg/kn/commands/revision_list.go:
```diff
@@ -36,6 +37,9 @@ func NewRevisionListCommand(p *KnParams) *cobra.Command {
 			namespace := cmd.Flag("namespace").Value.String()
+			if getAllNamespaces {
+				namespace = ""
+			}
```
How about
```
if getAllNamespaces {
  namespace := ""
} else {
  namespace := cmd.Flag("namespace").Value.String()
}
```
as `all-namespaces` take precedence over `namespace` so we check for it first.

same goes for service list

cppforlife replied:
@navidshaikh `:=` would define it inside the if block, so it wont be accessible below.

navidshaikh replied:
yess.. 👍

cppforlife commented:
@nak3 we should probably move this into KnParams or some other object (like NamespaceParams) that's injecting; otherwise, we would have to duplicate this functionality everywhere.

nak3 commented:
@cppforlife Thank you! Updated by adding the option into KnParams and adding the flag to revision,service command instead of list only.

cppforlife commented on pkg/kn/commands/revision.go:
```diff
@@ -24,6 +24,9 @@ func NewRevisionCommand(p *KnParams) *cobra.Command {
 		Short: "Revision command group",
 	}
 	revisionCmd.PersistentFlags().StringP("namespace", "n", "default", "Namespace to use.")
+	revisionCmd.PersistentFlags().BoolVar(&p.getAllNamespaces, "all-namespaces", false,
```
could we also share following 3 lines as well between commands?

nak3 replied:
Do you mean that `--all-namespace` should move to root level option like `--config` and `--kubeconfig`?
I put `--all-namespace` here to align with `--namespace`(`-n`). Should I share `namespace` between commands as well?

cppforlife replied:
@nak3 probably not global, but doesnt mean we cannot reuse this code so that we dont have duplicate flag definitions.

sixolet replied:
I'd love to see an AddNamespaceFlags fn and a GetNamespace(cmd) fn.

nak3 commented:
@cppforlife @sixolet  Thank you! Updated.

navidshaikh commented on pkg/kn/commands/namespaced.go:
```diff
@@ -0,0 +1,23 @@
+		"namespace",
+		"n",
+		"",
+		"If present, the namespace scope for this request",
```
IMO, we can re-phrase this to 

```
List the requested object(s) in given namespace.
```

navidshaikh commented on pkg/kn/commands/namespaced.go:
```diff
@@ -0,0 +1,23 @@
+		flags.Bool(
+			"all-namespaces",
+			false,
+			"If present, list the requested object(s) across all namespaces. Namespace in current context is ignored even if specified with --namespace",
```
Add the dot (.) at the end of statement.

navidshaikh commented on pkg/kn/commands/namespaced.go:
```diff
@@ -0,0 +1,49 @@
+		"namespace",
+		"n",
+		"",
+		"Namespace to use. (default \"default\")",
```
IMO, we can re-phrase this to
```
List the requested object(s) in given namespace.
```
makes the help text align with `all-namespaces` option.

cppforlife commented:
+1 for @navidshaikh suggestions

nak3 commented:
Thank you, updated the help messages.

cppforlife commented:
/ok-to-test

bbrowning commented:
Kicking these tests since a fix for these CI failures has since been merged.

/retest

nak3 commented:
rebased to solve confliction.

cppforlife commented:
/lgtm
/approve

nak3 commented:
/test pull-knative-client-unit-tests

nak3 commented:
The test was failed at TestServiceCreateImage. It looks like I need to rebase this PR again(?) Let's see the result of secod try.

```
I0408 17:44:11.183] --- FAIL: TestServiceCreateImage (0.00s)
I0408 17:44:11.183] panic: runtime error: invalid memory address or nil pointer dereference [recovered]
I0408 17:44:11.183] 	panic: runtime error: invalid memory address or nil pointer dereference
I0408 17:44:11.184] [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1827351]
I0408 17:44:11.184] 
I0408 17:44:11.184] goroutine 26 [running]:
I0408 17:44:11.184] testing.tRunner.func1(0xc0000efa00)
I0408 17:44:11.185] 	/usr/local/go/src/testing/testing.go:830 +0x68a
I0408 17:44:11.185] panic(0x1944e40, 0x29115e0)
I0408 17:44:11.185] 	/usr/local/go/src/runtime/panic.go:522 +0x1b5
I0408 17:44:11.185] github.com/knative/client/pkg/kn/commands.NewServiceCreateCommand.func1(0xc000492280, 0xc000289d40, 0x1, 0x3, 0x0, 0x0)
I0408 17:44:11.185] 	/go/src/github.com/knative/client/pkg/kn/commands/service_create.go:42 +0x1e1
I0408 17:44:11.186] github.com/spf13/cobra.(*Command).execute(0xc000492280, 0xc000289cb0, 0x3, 0x3, 0xc000492280, 0xc000289cb0)
I0408 17:44:11.186] 	/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:762 +0x528
I0408 17:44:11.186] github.com/spf13/cobra.(*Command).ExecuteC(0xc0002df680, 0x0, 0x0, 0x0)
I0408 17:44:11.187] 	/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:852 +0x3ee
I0408 17:44:11.187] github.com/spf13/cobra.(*Command).Execute(...)
I0408 17:44:11.187] 	/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:800
I0408 17:44:11.187] github.com/knative/client/pkg/kn/commands.fakeServiceCreate(0xc0000fc9b0, 0x5, 0x5, 0xc00019d760, 0xc00009b380, 0xc000482000, 0xc00019d730, 0x455cdd, 0x1eda418, 0xf)
I0408 17:44:11.188] 	/go/src/github.com/knative/client/pkg/kn/commands/service_create_test.go:59 +0x514
I0408 17:44:11.188] github.com/knative/client/pkg/kn/commands.TestServiceCreateImage(0xc0000efa00)
I0408 17:44:11.188] 	/go/src/github.com/knative/client/pkg/kn/commands/service_create_test.go:68 +0xc1
I0408 17:44:11.188] testing.tRunner(0xc0000efa00, 0x1b157f0)
I0408 17:44:11.189] 	/usr/local/go/src/testing/testing.go:865 +0x164
I0408 17:44:11.189] created by testing.(*T).Run
I0408 17:44:11.189] 	/usr/local/go/src/testing/testing.go:916 +0x69a
I0408 17:44:11.189] FAIL	github.com/knative/client/pkg/kn/commands	1.051s
```

nak3 commented:
@cppforlife @bbrowning @navidshaikh
Thank you. Sorry LGTM label was dropped as I rebased and added `AddNamespaceFlags()` to service create command - https://github.com/knative/client/pull/32/commits/3cad3dca9cc0157e2c689a8bf9eb9065d6ffbf19. Could you please take a look again?

cppforlife commented:
/lgtm
/approve
